### PR TITLE
CASMINST-5749: Improve validation of IUF Product Manifest

### PR DESCRIPTION
**Pull request by lathanm**
## Summary and Scope

Additional validation added to the IUF Product Manifest validator including the following:
1. Validate that files referenced in the IUF Product Manifest file actually exist in the release distribution directory. This probably requires that the validation be done late in the process of creating the release distribution, e.g. right before creating the tar file from the directory.
E.g., validate that the file paths specified under content.s3 actually exist.
E.g., validate that the files specified under content.nexus_repositories.yaml_path and content.nexus_blob_stores.yaml_path exist.
2. Validate that the Nexus repositories referenced in the iuf-product-manifest.yaml are hosted repositories. 
3. Validate that hosted repos are listed before group repos.
4. Validate that hosted/group repo references match.
5. Validate that there is actually content in the location specified by content.vcs.path, if specified
6. Validate that there is actually content in the location specified by content.rpms, if specified

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_

## Issues and Related PRs

https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-5749

* Future work required by (https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-6337)
* Documentation changes required in [issue id](issue link)

## Testing

Unit testing done with help of test data on local system

### Tested on:

  * Local development environment
  * hpc-shasta-os-cos-product-stream pipeline

### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

- Were new tests (or test issues/Jiras) created for this change? yes

## Pull Request Checklist

- [X] Version number(s) incremented, if applicable
- [X] Copyrights updated
- [X] License file intact
- [X] Target branch correct
- [X] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+ *  OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+package manifestDataValidation
+
+import (
+	"fmt"
+
+	mutils "github.com/Cray-HPE/cray-nls/src/api/models/iuf/mutils"
+	//v2yaml "gopkg.in/yaml.v2"
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:32]: Lets remove this commented stuff

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	exist := mutils.IsPathExist(file_path) // do we need error details??
+	if !exist {                            // if path is invalid
+		return fmt.Errorf("error in processing nexus repo file %v", file_path)
+	}
+
+	return nil
+}
+
+// Method to process nexus repo file and get hosted repo names
+func (vs *validators) processNexusRepoFile() error {
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:109]: validateNexusRepoFileContent() ?

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+	if err != nil {
+		return fmt.Errorf("failed to open Nexus Repository file: %v", err)
+	}
+
+	docs := mutils.SplitMultiYamlFile(nexusFile_contents)
+
+	skipFormats := []string{"docker", "helm"}
+
+	for _, d := range docs {
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:129]: for _, doc := docs {

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+		var nc interface{}
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:131]: nexusContentRaw

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+// Struct with list of validators
+type validators struct {
+	content           map[string]interface{}
+	nexusRepoFileName string
+	hostedRepoNames   []string
+}
+
+// Method to process s3 content, returns error in case of issues
+func (vs *validators) processS3() error {
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:67]: validateS3FilePath

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+	if len(temp_repo_names) > 0 {
+		return fmt.Errorf("Repo defined in host repo is not listed in group repo")
+	}
+
+	return nil
+}
+
+// Method to process nexus blob content, returns error in case of issues
+func (vs *validators) processNexusBlob() error {
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:188]: validateNexusBlobFilePath

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+	exist := mutils.IsPathExist(file_path) // do we need error details??
+	if !exist {                            // if path is invalid
+		return fmt.Errorf("error in processing nexus repo file %v", file_path)
+	}
+	return nil
+}
+
+// Method to process vcs content, return error in case of issues
+func (vs *validators) processVcs() error {
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:206]: validateVcsFilePath

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+	dir_path := vcs_map[VCS_PATH_KEY].(string)
+	empty := mutils.IsEmptyDirectory(dir_path) // do we need error details??
+	if empty {                                 // if path is invalid
+		return fmt.Errorf("error in processing vcs directory %v", dir_path)
+	}
+	return nil
+}
+
+// Method to process rpm content, return error in case of issues
+func (vs *validators) processRpm() error {
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:224]: validateRpmFilePath()

File: src/api/models/iuf/manifestDataValidation/manifest_data_validate.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+	for _, rpm := range rpm_array {
+		rpm_map := rpm.(map[string]interface{})
+
+		dir_path := rpm_map[RPM_PATH_KEY].(string)
+		empty := mutils.IsEmptyDirectory(dir_path) // do we need error details??
+		if empty {                                 // if path is invalid
+			return fmt.Errorf("error in processing rpm directory %v", dir_path)
+		}
+		repoName := rpm_map[REPO_NAME].(string)
```
Comment (arnab-muk-hpe) [inline src/api/models/iuf/manifestDataValidation/manifest_data_validate.go:242]: Lets move it out to create a separate function func (vs *validators) validateAllRpmsHosted() error{

**Review by arnab-muk-hpe — COMMENTED** [review]
@lathanm , added few comments, lets close it

**Review by arnab-muk-hpe — APPROVED** [review]
All test cases are passing. Approved,
```
go test ./...   
?       github.com/Cray-HPE/cray-nls    [no test files]
?       github.com/Cray-HPE/cray-nls/docs       [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/controllers        [no test files]
ok      github.com/Cray-HPE/cray-nls/src/api/argo-templates     (cached)
?       github.com/Cray-HPE/cray-nls/src/api/controllers/v1/misc        [no test files]
ok      github.com/Cray-HPE/cray-nls/src/api/controllers/v1/iuf (cached)
?       github.com/Cray-HPE/cray-nls/src/api/middlewares        [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/mocks/services     [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/routes     [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/services   [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/models/iuf/mutils  [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/models/iuf/schemaValidator [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/models/nls [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/models/nls/v1      [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/models/iuf [no test files]
?       github.com/Cray-HPE/cray-nls/src/api/models/iuf/manifestDataValidation  [no test files]
ok      github.com/Cray-HPE/cray-nls/src/api/controllers/v1/nls (cached)
```

**Review by atifsyedali — APPROVED** [review]

**Comment by atifsyedali** [comment]
@lathanm Remember that you will need corresponding version updates in https://github.com/Cray-HPE/cray-nls-charts/ and https://github.com/Cray-HPE/csm/

**Review by paulose-kuriakose — APPROVED** [review]

**Pull request merged**